source/benchmark_data.cc

```diff
@@ -749,219 +631,135 @@ Stream& operator<<(Stream &stream, const ChristensenBenchmark<dim> &christensen)
 
+  // Finite element values
+  FEValues<dim> fe_values(mapping,
+                          velocity.get_finite_element(),
+                          quadrature_formula,
+                          update_values|update_JxW_values);
 
-template <int dim>
-void ChristensenBenchmark<dim>::compute_local_global_squared_norms(
-  const typename DoFHandler<dim>::active_cell_iterator &cell,
-  AssemblyData::Benchmarks::Christensen::Scratch<dim>  &scratch,
-  AssemblyData::Benchmarks::Christensen::Copy          &data)
-{
-  // Reset local data
-  data.local_velocity_squared_norm        = 0.;
-  data.local_magnetic_field_squared_norm  = 0.;
-  data.local_discrete_volume              = 0.;
+  // Number of quadrature points
+  const unsigned int n_q_points = quadrature_formula.size();
 
-  // Initialize the finite element values
-  scratch.velocity_fe_values.reinit(cell);
+  // Vectors to stores the temperature gradients and normal vectors
+  // at the quadrature points
+  std::vector<Tensor<1, dim>> velocity_values(n_q_points);
 
   // Define the vector extractor
   const FEValuesExtractors::Vector  vector_extractor(0);
 
-  // Get the velocity values at each quadrature point
-  scratch.velocity_fe_values[vector_extractor].get_function_values(
-    velocity->solution,
-    scratch.velocity_values);
-
-  /*
-  // Repeat the steps for the magnetic flux
-  if (case_number != 0)
-  {
-    typename DoFHandler<dim>::active_cell_iterator
-    magnetic_flux_cell(&velocity->get_triangulation(),
-                        cell->level(),
-                        cell->index(),
-                        (magnetic_field->dof_handler).get());
+  // Integral value
+  double velocity_integral{0.0};
 
-    magnetic_flux_fe_values.reinit(magnetic_flux_cell);
-
-    magnetic_flux_fe_values[vector_extractor].get_function_values(
-      magnetic_field->solution,
-      magnetic_flux_values);
-  }
-  */
+  for (const auto &cell : velocity.get_dof_handler().active_cell_iterators())
+    if (cell->is_locally_owned())
+    {
+      // Initialize the finite element values
+      fe_values.reinit(cell);
+      // Get the velocity values at each quadrature point
+      fe_values[vector_extractor].get_function_values(velocity.solution,
+                                                      velocity_values);
+      // Numerical integration (Loop over all quadrature points)
+      for (unsigned int q = 0; q < n_q_points; ++q)
+      {
+        velocity_integral += velocity_values[q] * velocity_values[q] *
+                             fe_values.JxW(q);
+        discrete_volume += fe_values.JxW(q);
+      }
+    }
 
-  // Numerical integration (Loop over all quadrature points)
-  for (unsigned int q = 0; q < scratch.n_q_points; ++q)
+  // Gather the values of each processor
+  const Triangulation<dim>  &tria = velocity.get_triangulation();
+  const parallel::TriangulationBase<dim> *tria_ptr =
+      dynamic_cast<const parallel::TriangulationBase<dim> *>(&tria);
+  if (tria_ptr != nullptr)
```

Comment:
What is this for? The operation inside the scope is safe to use in serial, or not?

Reply:
The `dynamic_cast` checks at run time whether the pointer can be casted to a `parallel::TriangulationBase<dim>`. If not, the result is `nullptr` which means that the `Triangulation<dim>` are not a parallel one and, therefore, no MPI communication is necessary.

Reply:
This should allow all methods to be used in serial and in parallel. At one point it would be good to introduce the triangulation as a template parameter.